fqd_dss.c

```diff
@@ -33,10 +34,139 @@
 #include <poll.h>
 #include <pthread.h>
 
+#include <ck_fifo.h>
 #include <ck_pr.h>
 
 #define IN_READ_BUFFER_SIZE 1024*256
 
+static ck_fifo_spsc_t *work_queues;
+static uint32_t *work_queue_backlogs;
+static int *worker_thread_ids;
+static int worker_thread_count = 0;
+static pthread_t *worker_threads;
+static bool worker_thread_shutdown = false;
+
+struct incoming_message
+{
+  remote_data_client *client;
+  fq_msg *msg;
+};
+
+void *
+fqd_worker_thread(void *arg)
+{
+  struct incoming_message *m;
+  int tindex = *(int*) arg;
+
+  ck_fifo_spsc_t *q = &work_queues[tindex];
+  uint32_t *backlog = &work_queue_backlogs[tindex];
+
+  while (!worker_thread_shutdown) {
+    if (!CK_FIFO_SPSC_ISEMPTY(q)) {
+      if (ck_fifo_spsc_dequeue(q, &m)) {
+        if (m != NULL) {
+          ck_pr_dec_32(backlog);
+          fq_msg *copy = fq_msg_alloc_BLANK(m->msg->payload_len);
+          if (copy == NULL) {
+            continue;
+          }
+          memcpy(copy, m->msg, sizeof(fq_msg) + m->msg->payload_len);
+          /* reset the refcnt on the copy since the memcpy above will have overwritten it */
+          copy->refcnt = 1;
+          /* the copy will be freed as normal so eliminate cleanup_stack pointer */
+          copy->cleanup_stack = NULL;
+
+          /* we are done with the incoming message, deref */
+          fq_msg_deref(m->msg);
+
+          fqd_inject_message(m->client, copy);
+          free(m);
+        }
+      }
+    } else {
+      usleep(1);
+    }
+  }
+
+  /* drain the queue before exiting */
+  if (!CK_FIFO_SPSC_ISEMPTY(q)) {
+    ck_fifo_spsc_dequeue_lock(q);
+    if (ck_fifo_spsc_dequeue(q, &m)) {
+      if (m != NULL) {
+        ck_pr_dec_32(backlog);
+        fq_msg *copy = fq_msg_alloc_BLANK(m->msg->payload_len);
+        memcpy(copy, m->msg, sizeof(fq_msg) + m->msg->payload_len);
```

Comment:
It so that the incoming message (`m->msg`) can be plopped back onto the free list in the IO thread when it is deref'd.  The message allocated here is the real message that will live in memory queues until it is consumed.  

We could just simply not return it to the freelist in the IO thread and avoid this allocation but with any reasonable backlog in any queue the freelist would quickly become exhausted and throughput would suffer.  I can certainly test it to see how it does.

I can also segment the free list into sizes to avoid the large allocations that are happening now.